File: 구현/주문처리시스템/jeffrey/Order/Order.swift
//
//  Order.swift
//  ShoppingMall
//
//  Created by Mephrine on 1/8/25.
//

import Foundation

// MARK: - 주문 모듈
public final class Order: CustomStringConvertible {
    public enum Status {
        /// 주문 접수
        case pending
        /// 결제 완료
        case paid
        /// 취소됨
        case canceled
        /// 환불 완료
        case refunded
    }

    private let payment: any TransactionOperable
    private var orderedProducts: Products
    public private(set) var status: Status
    
    public init(cart: Cart, payment: any TransactionOperable) throws {
        guard cart.isEmpty == false else {
            throw CustomError.emptyCart
        }
        self.payment = payment
        self.orderedProducts = Products(products: cart.products)
        self.status = .pending
    }
    
    public func payment() async throws {
        guard status == .pending else {
            throw CustomError.paymentFailed
        }
        try await payment.processPayment(amount: orderedProducts.totalOrderAmount)
        self.status = .paid
    }
    
    public func cancel() throws {
        guard status == .pending else {
            throw CustomError.cancelFailed
        }
        self.status = .canceled
    }
    
    public func refund() async throws {
        guard status == .paid else {
            throw CustomError.refundFailed
        }
        try await payment.refund()
        self.status = .refunded
    }
    
    public var description: String {
        switch status {
        case .pending: "주문 접수"
        case .paid: "결제 완료."
        case .canceled: "취소됨"
        case .refunded: "환불 완료"
        }
    }
}

File: 구현/주문처리시스템/julie/Model/Order.kt
class Order(
    private val cart: Cart // 장바구니
) {
    var status: OrderStatus = OrderStatus.PENDING // 초기 주문 상태는 "주문 접수"
        private set
    private var totalAmount: Int = 0

    init {
        cart.validateNotEmpty() // 장바구니가 비어 있는지 확인
        totalAmount = cart.totalPrice() // 배송비 포함된 총 금액 계산
    }

    // 총 주문 금액 반환
    fun getTotalAmount(): Int = totalAmount

    // 주문 취소
    fun cancel() {
        if (status != OrderStatus.PENDING) throw CustomError.OrderNotAllowed // 결제 이전 상태에서만 취소 가능
        status = OrderStatus.CANCELLED
        println("주문이 취소되었습니다.")
    }

    // 결제 처리
    fun pay(paymentMethod: PaymentMethod) {
        if (status != OrderStatus.PENDING) throw CustomError.OrderNotAllowed // 결제 중복 방지
        paymentMethod.pay(totalAmount)
        status = OrderStatus.PAID
        println("결제가 완료되었습니다.")
    }

    // 환불 처리
    fun refund(paymentMethod: PaymentMethod) {
        if (status != OrderStatus.PAID) throw CustomError.RefundNotAllowed // 결제 완료 상태에서만 환불 가능
        paymentMethod.refund(totalAmount)
        status = OrderStatus.REFUNDED
        println("환불이 완료되었습니다.")
    }
}
